rest/replicator_test.go

```diff
@@ -333,6 +333,147 @@ func TestActiveReplicatorPullAttachments(t *testing.T) {
+
+	// Increase checkpoint persistence frequency for cross-node status verification
+	defer SetDefaultCheckpointInterval(50 * time.Millisecond)()
+
+	// Disable sequence batching for multi-RT tests (pending CBG-1000)
+	defer db.SuspendSequenceBatching()()
+
+	// Passive
+	tb2 := base.GetTestBucket(t)
+
+	rt2 := NewRestTester(t, &RestTesterConfig{
+		TestBucket: tb2,
+		DatabaseConfig: &DbConfig{
+			Users: map[string]*db.PrincipalConfig{
+				"alice": {
+					Password:         base.StringPtr("pass"),
+					ExplicitChannels: base.SetOf("alice"),
+				},
+			},
+		},
+		noAdminParty: true,
+	})
+	defer rt2.Close()
+
+	// Make rt2 listen on an actual HTTP port, so it can receive the blipsync request from rt1.
+	srv := httptest.NewServer(rt2.TestPublicHandler())
+	defer srv.Close()
+
+	passiveDBURL, err := url.Parse(srv.URL + "/db")
+	require.NoError(t, err)
+
+	// Add basic auth creds to target db URL
+	passiveDBURL.User = url.UserPassword("alice", "pass")
+
+	// Active
+	tb1 := base.GetTestBucket(t)
+
+	rt1 := NewRestTester(t, &RestTesterConfig{
+		TestBucket: tb1,
+		DatabaseConfig: &DbConfig{
+			Replications: map[string]*db.ReplicationConfig{
+				"repl1": {
+					Remote:                 passiveDBURL.String(),
+					Direction:              db.ActiveReplicatorTypePull,
+					Continuous:             true,
+					ConflictResolutionType: db.ConflictResolverCustom,
+					ConflictResolutionFn: `
+					function(conflict) {
+						var mergedDoc = new Object();
+						mergedDoc.source = "merged";
+
+						var mergedAttachments = new Object();
+						dst = conflict.RemoteDocument._attachments;
+						for (var key in dst) {
+							mergedAttachments[key] = dst[key];
+						}
+						src = conflict.LocalDocument._attachments;
+						for (var key in src) {
+							mergedAttachments[key] = src[key];
+						}
+						mergedDoc._attachments = mergedAttachments;
+
+						mergedDoc.channels = ["alice"];
+
+						return mergedDoc;
+					}`},
+			},
+		},
+	})
+	defer rt1.Close()
+
+	err = rt1.GetDatabase().SGReplicateMgr.StartReplications()
+	require.NoError(t, err)
+
+	rt1.waitForAssignedReplications(1)
+
+	docID := t.Name() + "doc1"
+	putDocResp := rt2.putDoc(docID, `{"channels":["alice"]}`)
+	require.True(t, putDocResp.Ok)
+	rev1 := putDocResp.Rev
+
+	// wait for the document originally written to rt2 to arrive at rt1
+	changesResults, err := rt1.WaitForChanges(1, "/db/_changes?since=0", "", true)
+	require.NoError(t, err)
+	require.Len(t, changesResults.Results, 1)
+	assert.Equal(t, docID, changesResults.Results[0].ID)
+	lastSeq := changesResults.Last_Seq.(string)
+
+	resp := rt1.SendAdminRequest(http.MethodPut, "/db/_replicationStatus/repl1?action=stop", "")
+	assertStatus(t, resp, http.StatusOK)
+
+	rt1.waitForReplicationStatus("repl1", db.ReplicationStateStopped)
+
+	putDocResp = rt1.putDoc(docID, `{"source":"rt1","`+db.BodyAttachments+`":{"localAtt.txt":{"data":"cmVtb3Rl"}},"channels":["alice"],"`+db.BodyRev+`":"`+rev1+`"}`)
```

Comment:
Does this work in the case where the local attachment wasn't new on the conflicting revision (i.e. the attachment had revpos/digest instead of data)?  Or is data always populated at this point, regardless of attachment history?  If so, do we need a test that verifies that the original metadata gets preserved by the eventual write?

Reply:
Yep, converted the test to use table driven tests to cover a bunch of scenarios.